File: android/src/main/java/com/rngooglemapsplus/MapUrlTileOverlayBuilder.kt
package com.rngooglemapsplus

import com.google.android.gms.maps.model.TileOverlay
import com.google.android.gms.maps.model.TileOverlayOptions
import com.google.android.gms.maps.model.UrlTileProvider
import java.net.URL

class MapUrlTileOverlayBuilder(
  private val mapErrorHandler: MapErrorHandler,
) {
  fun build(t: RNUrlTileOverlay): TileOverlayOptions {
    val provider =
      object : UrlTileProvider(
        t.tileSize.toInt(),
        t.tileSize.toInt(),
      ) {
        override fun getTileUrl(
          x: Int,
          y: Int,
          zoom: Int,
        ): URL? {
          val url =
            t.url
              .replace("{x}", x.toString())
              .replace("{y}", y.toString())
              .replace("{z}", zoom.toString())

          return try {
            URL(url)
          } catch (e: Exception) {
            mapErrorHandler.report(RNMapErrorCode.TILE_OVERLAY_FAILED, "tile url invalid: $url", e)
            null
          }
        }
      }

    val opts = TileOverlayOptions().tileProvider(provider)

    t.fadeIn?.let { opts.fadeIn(it) }
    t.zIndex?.let { opts.zIndex(it.toFloat()) }
    t.opacity?.let { opts.transparency(1f - it.toFloat()) }
    return opts
  }

  fun update(
    prev: RNUrlTileOverlay,
    next: RNUrlTileOverlay,
    overlay: TileOverlay,
  ) = onUi {
    if (prev.zIndex != next.zIndex) {
      overlay.zIndex = next.zIndex?.toFloat() ?: 0f
    }

    if (prev.opacity != next.opacity) {
      overlay.transparency = next.opacity?.let { 1f - it.toFloat() } ?: 0f
    }

    if (prev.fadeIn != next.fadeIn) {
      overlay.fadeIn = next.fadeIn ?: true
    }
  }
}

File: ios/MapUrlTileOverlayBuilder.swift
import GoogleMaps

class MapUrlTileOverlayBuilder {
  func build(_ t: RNUrlTileOverlay) -> GMSURLTileLayer {

    let constructor: GMSTileURLConstructor = { (x: UInt, y: UInt, zoom: UInt) in
      let urlString = t.url
        .replacingOccurrences(of: "{x}", with: "\(x)")
        .replacingOccurrences(of: "{y}", with: "\(y)")
        .replacingOccurrences(of: "{z}", with: "\(zoom)")
      return URL(string: urlString)
    }

    let layer = GMSURLTileLayer(urlConstructor: constructor)

    layer.tileSize = Int(t.tileSize)
    t.opacity.map { layer.opacity = Float($0) }
    t.zIndex.map { layer.zIndex = Int32($0) }
    t.fadeIn.map { layer.fadeIn = $0 }

    return layer
  }

  func update(
    _ prev: RNUrlTileOverlay,
    _ next: RNUrlTileOverlay,
    _ layer: GMSURLTileLayer
  ) {
    onMain {
      if prev.zIndex != next.zIndex {
        layer.zIndex = Int32(next.zIndex ?? 0)
      }

      if prev.opacity != next.opacity {
        layer.opacity = Float(next.opacity ?? 1)
      }

      if prev.fadeIn != next.fadeIn {
        layer.fadeIn = next.fadeIn ?? true
      }
    }
  }
}
